state/presence/pingbatcher.go

```diff
@@ -0,0 +1,267 @@
+func NewDeadPingBatcher(err error) *PingBatcher {
+	// we never start the loop, so the timeout doesn't matter.
+	pb := &PingBatcher{
+		pings:         nil,
+		pending:       make(map[string]slot),
+		flushInterval: time.Millisecond,
+		flushChan:     make(chan chan struct{}),
+	}
+	pb.tomb.Kill(err)
+	pb.tomb.Done()
+	return pb
+}
+
+// PingBatcher aggregates several pingers to update the database on a fixed schedule.
+type PingBatcher struct {
+	pings         *mgo.Collection
+	modelUUID     string
+	pending       map[string]slot
+	pingCount     uint64
+	flushInterval time.Duration
+	tomb          tomb.Tomb
+	pingChan      chan singlePing
+	flushChan     chan chan struct{}
+}
+
+// Start the worker loop.
+func (pb *PingBatcher) start() {
+	rand.Seed(time.Now().UnixNano())
+	go func() {
+		err := pb.loop()
+		cause := errors.Cause(err)
+		// tomb expects ErrDying or ErrStillAlive as
+		// exact values, so we need to log and unwrap
+		// the error first.
+		if err != nil && cause != tomb.ErrDying {
+			logger.Infof("ping batching loop failed: %v", err)
+		}
+		pb.tomb.Kill(cause)
+		pb.tomb.Done()
+	}()
+}
+
+// Kill is part of the worker.Worker interface.
+func (pb *PingBatcher) Kill() {
+	pb.tomb.Kill(nil)
+}
+
+// Wait returns when the Pinger has stopped, and returns the first error
+// it encountered.
+func (pb *PingBatcher) Wait() error {
+	return pb.tomb.Wait()
+}
+
+// Stop this PingBatcher, part of the extended Worker interface.
+func (pb *PingBatcher) Stop() error {
+	pb.tomb.Kill(nil)
+	err := pb.tomb.Wait()
+	return errors.Trace(err)
+}
+
+// nextSleep determines how long we should wait before flushing our state to the database.
+// We use a range of time around the requested 'flushInterval', so that we avoid having
+// all requests to the database happen at exactly the same time across machines.
+func (pb *PingBatcher) nextSleep() time.Duration {
+	sleepMin := float64(pb.flushInterval) * 0.8
+	sleepRange := float64(pb.flushInterval) * 0.4
+	offset := rand.Int63n(int64(sleepRange))
+	return time.Duration(int64(sleepMin) + offset)
+}
+
+func (pb *PingBatcher) loop() error {
+	flushTimeout := time.After(pb.nextSleep())
+	for {
+		select {
+		case <-pb.tomb.Dying():
+			return errors.Trace(tomb.ErrDying)
+		case <-flushTimeout:
+			// TODO (jam): 2017-06-21 I think I was wrong about channel selectivity.
+			// I think the problem was actually that we restart this timer every time we get a ping
+			// Switch back to a central loop and move this timeout
+			// outside of the loop, only creating it when it times out.
+			if err := pb.flush(); err != nil {
+				return errors.Trace(err)
+			}
+			flushTimeout = time.After(pb.nextSleep())
+		case singlePing := <-pb.pingChan:
+			pb.handlePing(singlePing)
+		case flushReq := <-pb.flushChan:
+			// Flush is requested synchronously.
+			// This way we know all pings have been handled. We will
+			// close the channel that was passed to us once we have
+			// finished flushing.
```

Comment:
This is a mechanism for both sides to synchronize. Side A wants a flush done, and wants to wait for that flush to complete. They send a channel, then wait for that channel to be closed. The other option would be to wait for a message on the channel.
However, waiting-for-a-message means we can block the loop() because the loop has to send the message. While 'closing a channel' can't ever block the loop() and still allows the callers to synchronize when they want to.